Add boundary conditions to Block and Domain classes

nilsdeppe opened the pull request:
## Proposed changes

- Have `Block` hold a `DirectionMap<std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>`. Currrently cannot enforce that these are set correctly because no domain creators support boundary conditions.
- Add support for boundary conditions to the `Domain` class and various helper functions that construct a `Domain`

The next step for supporting boundary conditions in the domain is adding support to the domain creators. At that point I think the elliptic and evolution codes can independently add support boundary conditions that are read from the input file.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add external boundary conditions to Block

nilsdeppe committed:
Add boundary conditions to Domain

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -19,15 +21,29 @@ Block<VolumeDim>::Block(
+        // We cannot enable this error message until boundary conditions are
+        // enabled everywhere in the code, unfortunately.
+        // ERROR("Couldn't find boundary condition for external boundary "
+        //       << direction);
```
I think you can already test the reverse case though: Make sure the user didn't pass a boundary condition for a direction that's actually internal

nilsvu commented on src/Domain/Domain.cpp:
```diff
@@ -27,13 +27,28 @@ template <size_t VolumeDim>
+        boundary_conditions) noexcept {
+  ASSERT(
+      boundary_conditions.size() == maps.size() or boundary_conditions.empty(),
+      "There must be either one boundary condition per block or none at all "
```
`...one set of boundary conditions per...` or similar (another occurrence below)

nilsvu commented on src/Domain/Block.hpp:
```diff
@@ -46,10 +47,15 @@ class Block {
+        DirectionMap<
+            VolumeDim,
+            std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>
+            external_boundary_conditions = {}) noexcept;
```
Do you need this to have a default for a particular reason? Once the consistency check in the implementation can be enabled I can't think of a case where one would not have to think about the set of boundary conditions when constructing a Block.

nilsdeppe replied:
Yes, once everything is updated we don't need the default, but I wasn't really feeling motivated to find all the defaults.

nilsvu commented on src/Domain/Domain.hpp:
```diff
@@ -53,9 +53,14 @@ class Domain {
+      std::vector<DirectionMap<
+          VolumeDim,
+          std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+          boundary_conditions = {}) noexcept;
```
same question about the necessity of this default. Also it would probably be good to add a note on the "specify all b.c. or none at all" requirement to the docs for this class.

nilsvu commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -310,7 +310,7 @@ std::vector<std::array<size_t, two_to_the(VolumeDim)>> corners_from_two_maps(
             vci_map1.local_corner_number();
         break;
-      } else {
+      } else {  // NOLINT(llvm-else-after-return)
```
llvm has a fair point here, don't you think? I understand if you don't want to refactor this code though. Instead of the NOLINT we can just merge and ignore the clang-tidy complaint.

nilsdeppe replied:
I don't understand what LLVM is trying to tell me. As far as I can tell it wants me to remove the `break;` but that actually changes the control flow.

nilsvu replied:
I think it wants you to delete the `else` scope, moving the code one indentation level down. The `else` doesn't really do anything here because the `break` in the `if` means the code below doesn't run anyway when the `if` is triggered.

nilsdeppe replied:
Ah, I'll try that and see if it works.

wthrowe replied:
Looks like a rename of readability-else-after-return, which is in our exclusion list.

nilsvu commented on tests/Unit/Domain/Test_Domain.cpp:
```diff
@@ -235,54 +271,86 @@ void test_1d_domains() {
+  std::vector<MapType> boundary_conditions{external_boundaries.size()};
+  for (size_t i = 0; i < external_boundaries.size(); ++i) {
+    const auto& external_boundaries_of_block = external_boundaries[i];
+    MapType t{};
```
t -> boundary_conditions_i ?

nilsvu reviewed (changes requested):
Two design questions that we can also discuss in a call if you prefer:

1. I missed to ask in #2750: Doesn't the `domain::BoundaryConditions::BoundaryCondition` base class need the list of registrars as template parameter? If I understand correctly it needs to build the `creatable_classes` list so it can be factory-created from domain creators.
2. Do you have an idea how to handle cases where boundary conditions are not needed, e.g. for things like `ExportCoordinates` or when we create an `ApplyObservers`-type executable or do interpolations between domains?

nilsdeppe commented:
- This doesn't work without templating the domain on the system or metavariables, which I don't think we should do. Instead, each system defines another base class that is used for the factor. The interval domain creator will have tags:

  ```cpp
  template <typename BoundaryConditionsBase>
  struct UpperBoundaryCondition {
    static std::string name() noexcept { return "UpperBoundary"; }
    static constexpr Options::String help =
        "Options for the boundary condition applied at the upper boundary.";
    using type = std::unique_ptr<BoundaryConditionsBase>;
    using group = BoundaryConditions;
  };
  template <typename BoundaryConditionsBase>
  struct LowerBoundaryCondition {
    static std::string name() noexcept { return "LowerBoundary"; }
    static constexpr Options::String help =
        "Options for the boundary condition applied at the lower boundary.";
    using type = std::unique_ptr<BoundaryConditionsBase>;
    using group = BoundaryConditions;
  };

  using basic_options =
      tmpl::list<LowerBound, UpperBound, IsPeriodicIn, InitialRefinement,
                 InitialGridPoints, TimeDependence>;

  template <typename Metavariables>
  using options = tmpl::conditional_t<
      domain::BoundaryConditions::has_boundary_conditions_base_v<
          typename Metavariables::system>,
      tmpl::push_back<
          basic_options,
          LowerBoundaryCondition<
              domain::BoundaryConditions::get_boundary_conditions_base<
                  typename Metavariables::system>>,
          UpperBoundaryCondition<
              domain::BoundaryConditions::get_boundary_conditions_base<
                  typename Metavariables::system>>>,
      basic_options>;
  ```

  Basically, the system has a `using boundary_conditions_base =...` type alias that points to the base class that is factory-createable. This base class much inherit from `domain::BoundaryConditions::BoundaryCondition`
- They will need to set up some empty system with an a `None` boundary condition. We could make specifying boundary conditions optional, but that seems fragile. As per `ApplyObservers`: I think that should probably have access to the boundary conditions. One may conceivably want to observe the boundary condition terms being added to the time derivatives.

nilsvu commented:
I see, thanks for explaining. I agree we shouldn't template the domain on the metavariables or the system, that seems bad. We _could_ template the domain creators on the boundary condition base class when we update them to support boundary conditions. I'm suggesting this because I don't like the pattern of "reaching up" to the metavariables to retrieve a special type alias and much prefer the pattern of "passing down" such information from the metavariables. If the "reaching up" pattern seems unavoidable then I would suggest to make the type alias part of the metavariables, not the system. The reason is that multiple executables may use the same system but want to specify a different set of possible boundary conditions. For example, a `SolveBBH` executable offers the choice of apparent-horizon boundary conditions whereas a `SolveBNS` executable does not. Similarly, a `SolveThermalNoise` executable offers a `LaserBeam` boundary condition but a `SolveBentMaterial` executable does not, but both solve an elasticity system.

nilsdeppe commented:
What I don't like about templating all of the domain creator on the boundary conditions is that this means either more code in hpp files (bad for compile time) or _a lot_ of `tpp` files that require many explicit instantiations (not super user friendly, I think). I think the benefit of adding the boundary conditions as a template parameter to the domain creators is one could easily solve different systems on different domains. This _might_ be of interest when we solve the Boltzmann equation, but I think there are way too many other problems to figure out before then to make that a compelling argument for making the code harder to use now.

I intentionally didn't add the type alias to the `metavariables` since boundary conditions are valid for specific systems only. I also find having lots of different options listed as separate type aliases in the metavariables makes the code impossible to understand. From your list: `SolveBBH` and `SolveBNS` are solving different physical systems, one has matter one does not. You could template the `System` class on boundary conditions if you want to reduce code, but I think they should not be treated as the same system. The second example: you are trying to restrict the boundary conditions for the user. While reasonable, I don't think we should make the evolution code harder to use because of this. Further more, if I can specify the domain, initial conditions, and boundary conditions in the input file, doesn't that mean you can combine both the executables you listed into a single executable? This is what the evolution code is going to do, and then it is up to the user to specify a reasonable physical setup.

In short, I like the current design, if you want to restrict what boundary conditions are available I think you should template the system on boundary conditions.

nilsvu commented:
+1 to keeping domain creator code in cpp files. Boundary conditions should disrupt the domain code as little as possible. The particular base class is only needed for the option-creation process, and I also don't see another way of injecting compile-time information into the option-creation without going though the metavariables.

About metavars vs system: We might end up with exactly one executable per system, but I don't think we know that for sure right now. Seems too early to make that call..

nilsdeppe commented:
Yes, but at least in the evolution there are many systems (e.g. Newtonian Euler) where the evolution larger differs only by the domain, boundary conditions, and initial condition. We will likely have separate executables for BBH and single BH, but that's more so because of the AH finders and control systems rather than the system/domain. Does templating the system on the boundary conditions work for your use case?

nilsvu commented:
Sure it works to template the system on the boundary conditions, but then all it does is receive the template parameter and expose it as an alias, which seems silly 😄 Perhaps I could make the system use all boundary conditions by default and allow a template parameter to restrict the set.

nilsdeppe commented:
Okay :)

nilsdeppe committed:
Add llvm-else-after-return to clang-tidy ignore list

nilsdeppe commented:
pushed fixups and a new commit that adds the `llvm-...` to the clang-tidy ignore list

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -28,6 +28,11 @@ Block<VolumeDim>::Block(
       neighbors_(std::move(neighbors)) {
+  ASSERT(external_boundary_conditions.empty() or
+             external_boundary_conditions.size() ==
+                 2 * VolumeDim - neighbors_.size(),
```
This assert doesn't catch the case where `external_boundary_conditions` has the correct number of entries, but at the wrong directions. I suggest to test that, perhaps wrap the loop in `#ifdef SPECTRE_DEBUG` if necessary

nilsvu commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -1303,7 +1318,8 @@ ShellWedges Options::create_from_yaml<ShellWedges>::create<void>(
   if (which_wedges == "All") {
     return ShellWedges::All;
-  } else if (which_wedges == "FourOnEquator") {
+  } else if (which_wedges ==  // NOLINT(llvm-else-after-return)
```
here's a left-over NOLINT

nilsvu reviewed (changes requested):
Please squash, including these

nilsdeppe commented:
Rebased and squashed. I added an `ASSERT` inside the existing for loop that checks that either there are no boundary conditions at all or the current external direction has a boundary condition specified. Thanks for the review :)

kidder commented on src/Domain/Block.cpp:
```diff
@@ -98,8 +125,21 @@ std::ostream& operator<<(std::ostream& os,
+  // number of external boundaries set.
   return lhs.id() == rhs.id() and lhs.neighbors() == rhs.neighbors() and
          lhs.external_boundaries() == rhs.external_boundaries() and
+         alg::all_of(rhs.external_boundaries(),
```
I don't see how this tests anything about the equivalence of Blocks since `lhs.external_boundaries() == rhs.external_boundaries()`.  It will only fail for an invalid Block which should be caught at construction...

nilsdeppe replied:
heh, yea I think you're right. The other test I can add is that the `typeid` of the dereferenced base class pointers are the same to make sure that at the least the boundary conditions are of the same type. Thoughts?

kidder replied:
seems reasonable

nilsdeppe commented:
Rebased and squashed in the additional comparisons

kidder commented:
@nilsleiffischer please look at

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -98,8 +125,19 @@ std::ostream& operator<<(std::ostream& os,
+               return lhs.external_boundary_conditions().contains(dir) and
+                      rhs.external_boundary_conditions().contains(dir) and
+                      typeid(lhs.external_boundary_conditions().at(dir)) ==
+                          typeid(rhs.external_boundary_conditions().at(dir));
```
I believe you have to `#include <typeinfo>` to use `typeid`. While you're at it you might as well add the includes for `Direction`, `DirectionMap`, `<memory>`, `BlockNeighbor`, `<utility>`

nilsvu commented on tests/Unit/Domain/Test_Block.cpp:
```diff
@@ -26,19 +27,72 @@
 #include "Domain/Structure/OrientationMap.hpp"
 #include "Framework/TestHelpers.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Utilities/CloneUniquePtrs.hpp"
```
I think you don't actually use `clone_unique_ptrs`

nilsvu commented on src/Domain/Domain.hpp:
```diff
@@ -52,10 +52,17 @@ class Domain {
+      std::vector<std::unique_ptr<domain::CoordinateMapBase<
+          Frame::Logical, Frame::Inertial, VolumeDim>>>
+          maps,
+      std::vector<DirectionMap<
```
While squashing, add include for `DirectionMap` and `domain::BoundaryConditions::BoundaryCondition`.

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -19,15 +21,39 @@ Block<VolumeDim>::Block(
+             "Specifying boundary conditions but the external direction "
+                 << direction << " in block " << id_
+                 << " does not have a specified boundary condition.");
+      if (not external_boundary_conditions.contains(direction)) {
```
(optional) I find it easier to follow control flow with fewer boolean inversions, so I suggest to swap the `if` and `else` clauses here.

nilsvu reviewed (commented):
Please squash these in directly

nilsdeppe commented:
rebased and squashed in changes. I removed `<ostream>` from `Block.hpp`, replacing it with `<iosfwd>` since that's all it needs. Thanks for the reviews!
